beamer/check/commands.py

```diff
@@ -723,3 +723,89 @@ def initiate_challenges(
             json.dump(data, f, indent=4)
 
     log.info("All challenges initiated succesfully")
+
+
+@check.command("verify-challenges")
+@click.option(
+    "--rpc-file",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to the RPC config file.",
+)
+@click.option(
+    "--artifacts-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="The directory containing deployment artifacts.",
+)
+@click.option(
+    "--abi-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="Path to the directory with contract ABIs.",
+)
+@click.argument(
+    "file", type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path)
+)
+def verify_challenges(rpc_file: Path, artifacts_dir: Path, abi_dir: Path, file: Path) -> None:
+    """Verify challenges stored in FILE."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    abi_manager = ABIManager(abi_dir)
+    rpc_info = beamer.util.load_rpc_info(rpc_file)
+    ctx = Context(
+        abi_manager=abi_manager, account=None, artifacts_dir=artifacts_dir, rpc_info=rpc_info
+    )
+
+    with file.open("rt") as f:
+        data = json.load(f)
+
+    challenges = apischema.deserialize(list[Challenge], data)
+
+    errors = 0
+    for challenge in challenges:
+        # First check the finalization timestamp.
+        if (
+            challenge.finalization_timestamp is None
+            or datetime.utcnow() <= challenge.finalization_timestamp
+        ):
+            log.error(
+                "Challenge not yet finalized, skipping",
+                txhash=challenge.challenge_claim_txhash,
+                request_chain=challenge.request_chain,
+                fill_chain=challenge.fill_chain,
+            )
+            errors += 1
+            continue
+
+        # Check whether L1 resolution was performed correctly.
+        request_manager = _obtain_request_manager(ctx, challenge.request_chain)
+        claim = request_manager.functions.claims(challenge.claim_id).call()
+        request = request_manager.functions.requests(challenge.request_id).call()
+        if (
+            request.withdrawClaimId == challenge.claim_id
```

Comment:
Unfortunately, this does improve the situation only a bit. It makes sure that it was withdrawn by this very claimId but not before or after the resolution, or IOW it doesn't state who the beneficiary was. However this is not so easy to find out. So I would be fine with this tradeoff

Reply:
> It makes sure that it was withdrawn by this very claimId but not before or after the resolution

Actually, due to the fact that `request.filler` is only set by `resolveRequest`, it's guaranteed that L1 resolution occurred if that condition is true. However, ....

> IOW it doesn't state who the beneficiary was.

... is definitely true, we don't have an easy way to check that the beneficiary was the claimer.
The only thing that comes to my mind is looking for the exact withdrawal event and comparing the addresses, which would be :woozy_face: .

Here I'm not so much concerned about the complexity of event-based checking, we have all the infrastructure to make that relatively straightforward. Rather, I'm more concerned about the overall "cost" of the command i.e. now verification is quite quick and costs nothing and when I weigh that against the event-based checking... eh, I think I prefer more the quick-and-cheap-to-verify (also in presence of many challenges in one file).